### DISCUSSION TITLE
Early Hints

### DISCUSSION THREAD

**Discussion by davidbarratt** [vercel/next.js#36089] [Ideas]
### Describe the feature you'd like to request

Some CDNs like [Cloudflare](https://developers.cloudflare.com/cache/about/early-hints/) support the [Early Hints](https://httpwg.org/specs/rfc8297.html) web standard.

This can dramatically improve the page loading speed as the CDN can hint to the browser the render-blocking resources (CSS, Fonts, etc.) that need to be downloaded.

### Describe the solution you'd like

Next.js could add the appropriate [Link](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Link) headers to any HTML response with [preload](https://developer.mozilla.org/en-US/docs/Web/HTML/Link_types/preload) for any stylesheets that Next.js is going to add the `<head>`. Doing so will allow CDNs to cache these headers. On the next request, they can return the headers early while the CDN waits for the complete response.

### Describe alternatives you've considered

I've considered implementing this ourselves within Next.js, but it is difficult (if not impossible?) to know the URLs to the CSS files from _within_ Next.js itself.

**Comment by awareness481** [comment]
No browsers currently support Early hints (afaik Chrome only did as an experiment during versions 94 - 98) and you'd probably use either server streaming or early hints.
It's probably years away from browser adoption but I would love to hear thoughts from the Next.js team.

**Reply by fabb** [reply]
It’s available in Chrome 103

**Reply by benmccann** [reply]
It's also available in Firefox 117 and Edge 103

**Comment by pierophp** [comment]
Early Hints was shipped with Chrome 103.

https://blog.cloudflare.com/early-hints-performance/

Maybe is time to review this

**Reply by fabb** [reply]
@timneutkens are there plans to add support for early hints? (Please also consider support when using a custom express server)

**Reply by BleedingDev** [reply]
Now with Node.js 18.11 we can do it with with custom server. Native support would be great tho! :)

**Reply by childrentime** [reply]
how to use it with custom server?

**Comment by pierophp** [comment]
As a workaround, if you have a Cloudflare Worker in front of your site, you could parse the response and add the header.

I made a simple example of that:

https://gist.github.com/pierophp/b790ae748bb8e810abca4336424ae651

It added the header correctly, but it seems Cloudflare is not giving the early hints. I still don't know why.

I found these threads:

https://community.cloudflare.com/t/early-hints-need-more-data-before-switching-over/342888
https://community.cloudflare.com/t/early-hints-dont-work-on-cloudflare/394263

**Reply by derryl** [reply]
FYI Cloudflare now officially supports [Early Hints](https://developers.cloudflare.com/cache/about/early-hints/) (at least for Cloudflare Pages – I'm not sure about standalone worker functions.)

**Comment by YoavRheims** [comment]
This is a must-have, no question about it.

**Reply by revskill10** [reply]
I have one question: WHy ?

**Comment by YoavRheims** [comment]
Cool! Yeah, native support would be great.

**Comment by eric-burel** [comment]
Isn't the new approach of separating the head tags in their own files (https://beta.nextjs.org/docs/routing/pages-and-layouts#modifying-head) a first step towards this? Because Next.js can evaluate this file without having to evaluate the actual page

**Comment by marufmax** [comment]
waiting for this to release

**Comment by chris-ng-1987** [comment]
the feature is well supported in most browser already, any update to support it?

https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/103

**Comment by childrentime** [comment]
This should be a issue......

**Comment by childrentime** [comment]
https://github.com/vercel/next.js/issues/66980

I attempted to add this feature in a custom server with Next.js. However, the only issue is that after using res.flushHeaders or res.write('') to send the response headers, Next.js does not send any further response.

**Comment by amitkoth** [comment]
Really need control of how CSS and JS are injected to automate early hints generation on Cloudflare, as noted above, and here:

https://developers.cloudflare.com/pages/configuration/early-hints/

**Comment by childrentime** [comment]
nuxtjs support it😅😅😅

**Comment by childrentime** [comment]
well. There is a way to access this purpose, but you need to enable app dir. you can refer to this
https://paretojs.netlify.app/blog/early-hints.html

**Comment by joostmeijles** [comment]
Could NextJS middleware somehow be used to send the Early Hints response header?

**Comment by searchstefano** [comment]
If you may be interested I have opened an issues to bring this functionality into Nextjs 15: https://github.com/vercel/next.js/issues/71722

**Comment by DavideSegullo** [comment]
I agree now it seems to be more supported and stable:

https://developer.chrome.com/docs/web-platform/early-hints

**Reply by thaophamvan** [reply]
Is there any example implement early hints in nextjs

**Comment by amitkoth** [comment]
Any update on this? All Cloudflare needs is the right response format and it _automatically_ generates early hints:

https://developers.cloudflare.com/pages/configuration/early-hints/#2-automatic-link-header-generation

**Comment by Dev-Ls-07** [comment]
eu já tive um pouco de experiência com front-end (html, css e javascript), como é você passar disso para frameworks como angular, bootstrap e outras coisas como node e react?

**Comment by PanSzelescik** [comment]
It would be cool if next.js will generate Link headers for Cloudflare's Early Hints: [https://developers.cloudflare.com/cache/advanced-configuration/early-hints/#generate-early-hints](https://developers.cloudflare.com/cache/advanced-configuration/early-hints/#generate-early-hints)
